### ISSUE TITLE
Support for primitive type names in schema enum types

### ISSUE THREAD

**Issue by theBradfo** [apollographql/apollo-kotlin#4023]
**Summary**
We have an existing schema defined enum that uses primitive types as values.  This generates code all fine and well through Apollo, but the kotlin compiler breaks on the `rawValue` argument to the enum.

**Version**
3.2.0

**Description**
We have an existing enum defined in the schema:
```
enum SomethingPrimitiveType {
  Float

  Long

  String

  Timestamp

  Boolean

  Bytes

  Record
}
```
And Apollo generates the kotlin source for this as `SomethingPrimitiveType.kt`: 
```
public enum class SomethingPrimitiveType(
  public val rawValue: String,
) {
  Float("Float"),
  Long("Long"),
  String("String"),
  Timestamp("Timestamp"),
  Boolean("Boolean"),
  Bytes("Bytes"),
  Record("Record"),
  /**
   * Auto generated constant for unknown enum values
   */
  UNKNOWN__("UNKNOWN__"),
  ;

  public companion object {
    public val type: EnumType = EnumType("SomethingPrimitiveType")

    public fun safeValueOf(rawValue: String): SomethingPrimitiveType =
        values().find { it.rawValue == rawValue } ?: UNKNOWN__

    /**
     * Returns all [SomethingPrimitiveType] known at compile time
     */
    public fun knownValues(): Array<SomethingPrimitiveType> = arrayOf(
      Float,
      Long,
      String,
      Timestamp,
      Boolean,
      Bytes,
      Record)
  }
}
```
The kotlin compiler chokes on this due to the `rawValue` argument being of type `String`, which in this case it sees as the `SomethingPrimitiveType.String` enum value.  Can we generate this `rawValue` argument as `kotlin.String` explicitly?

**Comment by martinbonnin** [comment]
That's a though one... First thing that comes to mind is adding a Gradle option that would [alias the types](https://square.github.io/kotlinpoet/1.x/kotlinpoet/kotlinpoet/com.squareup.kotlinpoet/-file-spec/-builder/add-aliased-import.html):

```
apollo {
   addAliasImport("kotlin.String", "MyString")
}
```

But that would change all the `kotlin.String` in all of the codegen. Alternatively we could start thinking about good extension points in the compiler so that the `FileSpec.Builder` are ultimately configurable and expose that as API

**Comment by theBradfo** [comment]
Yeah, its likely an edge case as well and we're only seeing this b/c we have this existing, massive schema 😅.  For now we can ignore this type as none of the operations we're converting to use Apollo are using it yet, but that day will likely come at some point!

**Comment by Judas** [comment]
I also have this problem with a Language codes enum.
Values for **as**, **in** & **is** are well quoted in the generated enum class but they are not in the **knownValues()** array

```
public enum class Language(
  public val rawValue: String,
) {
  ... 

  /**
   * Arabic (ar)
   */
  ar("ar"),
  /**
   * Assamese (as)
   */
  `as`("as"),
 
  ... 

  /**
   * Inupiaq (ik)
   */
  ik("ik"),
  /**
   * Indonesian (in)
   */
  `in`("in"),
  /**
   * Ido (io)
   */
  io("io"),
  /**
   * Icelandic (is)
   */
  `is`("is"),
  
  /**
   * Auto generated constant for unknown enum values
   */
  UNKNOWN__("UNKNOWN__"),
  ;

  public companion object {
    public val type: EnumType = EnumType("Language")

    public fun safeValueOf(rawValue: String): Language = values().find { it.rawValue == rawValue }
        ?: UNKNOWN__

    /**
     * Returns all [Language] known at compile time
     */
    public fun knownValues(): Array<Language> = arrayOf(
      ...
      ar,
      as, // breaks compilation
      ...
      ik,
      in, // breaks compilation
      io,
      is, // breaks compilation
     ...
    )
  }
}
```

**Comment by martinbonnin** [comment]
@Judas good catch! This one might be a bit easier to fix as KotlinPoet supports adding backticks. I'd say changing a `%L` to a `%N` somewhere should to the trick. Maybe [here](https://github.com/apollographql/apollo-kotlin/blob/0c06bb3a78c042c4c7afcbca492e3fe181e6c7e8/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/file/EnumAsEnumBuilder.kt#L83). Do you want to give it a shot and open a PR?

**Comment by martinbonnin** [comment]
@theBradfo turns out this is something that we can probably fix at the KotlinPoet level: https://github.com/square/kotlinpoet/issues/1234

@Judas since I was looking at this, I took a stab at the `%N` thing and it seems to work 🎉 : https://github.com/apollographql/apollo-kotlin/pull/4035

**Comment by kaganovych** [comment]
I faced the same issue with `in` and `out` from the generated enum. Was it fixed in the 3.3.0 version?

Also, I have kind of the same issue, but now it's with conflicting names: 

```
public enum class ListCarsSortingFields(
  public val rawValue: String,
) {
  color("color"),
  type("type"),
  year("year"),
  /**
   * Auto generated constant for unknown enum values
   */
  UNKNOWN__("UNKNOWN__"),
  ;

  public companion object {
    public val type: EnumType = EnumType("ListCarsSortingFields")

    public fun safeValueOf(rawValue: String): ListCarsSortingFields =
        values().find { it.rawValue == rawValue } ?: UNKNOWN__

    /**
     * Returns all [ListCarsSortingFields] known at compile time
     */
    public fun knownValues(): Array<ListCarsSortingFields> = arrayOf(
      color,
      type,
      year)
  }
}
```

Here in `knownValues()`, the `type` field is taken from the `companion object,` not from the enum itself, and it doesn't compile.

**Comment by BoD** [comment]
In v3.3.1 which was just released, you can now add the `@targetName` directive on enum values, like so:

```graphql
# Inside an extra.graphqls file to put next to your schema.graphqls file

extend enum ListCarsSortingFields {
  type @targetName(name: "byType")
}
```
This will make the generated name be `byType` instead of `type`.

Note: this is an experimental feature.

**Comment by martinbonnin** [comment]
This is fixed in 3.3.2
